File: KWayMerge/src/main/python/solution373.py
from typing import List
import heapq


class Solution:
    def kSmallestPairs(
        self, nums1: List[int], nums2: List[int], k: int
    ) -> List[List[int]]:

        n1 = len(nums1)
        n2 = len(nums2)
        rever = False

        if n1 > n2:
            n1, n2 = n2, n1
            nums1, nums2 = nums2, nums1
            rever = True

        hq = []
        ans = []

        for i in range(min(n1, k)):
            heapq.heappush(hq, [nums1[i] + nums2[0], i, 0])

        while len(ans) < k and hq:
            _, i, j = heapq.heappop(hq)
            ans.append([nums1[i], nums2[j]] if not rever else [nums2[j], nums1[i]])

            if j + 1 < n2:
                heapq.heappush(hq, [nums1[i] + nums2[j + 1], i, j + 1])
        return ans


s = Solution()
nums1 = [6, 7, 11]
nums2 = [2, 4, 6]
k = 3
res = s.kSmallestPairs(nums1, nums2, k)

File: KWayMerge/src/main/java/Solution373.java
import java.util.*;

class Solution373 {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {

        int n1 = nums1.length;
        int n2 = nums2.length;

        List<List<Integer>> ans = new ArrayList<>();

        PriorityQueue<int[]> pq = new PriorityQueue<>(
                new Comparator<int[]>() {
                    @Override
                    public int compare(int[] o1, int[] o2) {
                        return o1[0] - o2[0];
                    }

                });

        for (int i = 0; i < n1; i++) {
            pq.add(new int[] { nums1[i] + nums2[0], i, 0 });
        }

        while (ans.size() < k && !pq.isEmpty()) {
            int[] l = pq.poll();
            int i = l[1];
            int j = l[2];

            List<Integer> content = new ArrayList<>();

            content.add(nums1[i]);
            content.add(nums2[j]);

            ans.add(content);

            if (j + 1 < n2) {
                pq.add(new int[] { nums1[i] + nums2[j + 1], i, j + 1 });
            }

        }
        return ans;

    }
}